NiBAx/plugins/harmonizationplugin/harmonization.py

```diff
@@ -0,0 +1,169 @@
+            MUSEDictDataFrame= self.datamodel.GetMUSEDictDataFrame()
+            muse_mappings = self.datamodel.GetDerivedMUSEMap()
+            for ROI in MUSEDictDataFrame[MUSEDictDataFrame['ROI_LEVEL']=='DERIVED']['ROI_INDEX']:
+                single_ROIs = muse_mappings.loc[ROI].replace('NaN',np.nan).dropna().astype(np.float)
+                single_ROIs = ['H_MUSE_Volume_%0d' % x for x in single_ROIs]
+                muse['H_MUSE_Volume_%d' % ROI] = muse[single_ROIs].sum(axis=1,skipna=False)
+            muse.drop(columns=['H_MUSE_Volume_702'], inplace=True)
+
+        start_index = len(self.datamodel.harmonization_model['SITE_labels'])
+        sex_icv_effect = np.dot(muse[['Sex','DLICV_baseline']].copy(), self.datamodel.harmonization_model['B_hat'][start_index:(start_index+2),:])
+        ROIs_ICV_Sex_Residuals = ['RES_ICV_Sex_' + x for x in self.datamodel.harmonization_model['ROIs']]
+        muse.loc[:,ROIs_ICV_Sex_Residuals] = muse[['H_' + x for x in self.datamodel.harmonization_model['ROIs']]].values - sex_icv_effect
+
+        muse.loc[:,'Sex'] = muse['Sex'].map({1:'M',0:'F'})
+        ROIs_Residuals = ['RES_' + x for x in self.datamodel.harmonization_model['ROIs']]
+        RAW_Residuals = ['RAW_RES_' + x for x in self.datamodel.harmonization_model['ROIs']]
+        muse.loc[:,ROIs_Residuals] = bayes_data-stand_mean
+        muse.loc[:,RAW_Residuals] = Raw_ROIs_Residuals
+        print('Harmonization done.')
+
+        return muse, parameters
+
+
+    def AddHarmonizedMUSE(self, muse):
+        print('Saving modified data to pickle file...')
+
+        muse.set_index(self.datamodel.data.index,inplace=True)
+
+        ROI_list = list(self.datamodel.harmonization_model['ROIs'])
+        if ('MUSE_Volume_301' not in ROI_list):
+            logger.info('No derived volumes in model')
+            MUSEDictDataFrame= self.datamodel.GetMUSEDictDataFrame()
+            Derived_numbers = list(MUSEDictDataFrame[MUSEDictDataFrame['ROI_LEVEL']=='DERIVED']['ROI_INDEX'])
+            Derived_MUSE_Volumes = list('MUSE_Volume_' + str(x) for x in Derived_numbers)
+            ROI_list = ROI_list + Derived_MUSE_Volumes
+            ROI_list.remove('MUSE_Volume_702')
+        else:
+            logger.info('Model includes derived volumes')
+        H_ROIs = list('H_' + str(x) for x in ROI_list)
+        ROIs_ICV_Sex_Residuals = ['RES_ICV_Sex_' + x for x in self.datamodel.harmonization_model['ROIs']]
+        ROIs_Residuals = ['RES_' + x for x in self.datamodel.harmonization_model['ROIs']]
+        RAW_Residuals = ['RAW_RES_' + x for x in self.datamodel.harmonization_model['ROIs']]
+        if ('H_MUSE_Volume_47' not in self.datamodel.data.keys()):
+            self.datamodel.data.loc[:,H_ROIs] = muse[H_ROIs]
+        self.datamodel.data.loc[:,ROIs_ICV_Sex_Residuals] = muse[ROIs_ICV_Sex_Residuals]
+        self.datamodel.data.loc[:,ROIs_Residuals] = muse[ROIs_Residuals]
+        self.datamodel.data.loc[:,RAW_Residuals] = muse[RAW_Residuals]
```

Comment:
These data updates should happen in datamodel class.

Reply:
@ashishsingh18 Agreed. This will require a generic `AddDataToDatamodel` in `datamodel` that we don't have yet. I propose to implement this in a separate PR.